docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,627 @@
+
+    m = torch.jit.script(inc)
+    print(inc((1,2)))
+
+Running the above codes yield the following scripting error. The remedy is to add from typing import Tuple.
+
+.. testoutput::
+
+    File "test-tuple.py", line 5, in <module>
+        def inc(x: Tuple[int, int]):
+    NameError: name 'Tuple' is not defined
+
+Nominal Types
+^^^^^^^^^^^^^
+
+Nominal TorchScript types are Python classes. They are called nominal because these types are declared with a custom
+name and are compared using class names. Nominal classes are further classified into the following categories:
+
+::
+
+    TSNominalType := TSBuiltinClasses | TSCustomClass | TSEnum
+
+Among them, TSCustomClass and TSEnum must be compilable to TorchScript IR (as enforced by the type-checker).
+
+Builtin Class
+^^^^^^^^^^^^^
+
+Builtin nominal types are Python classes whose semantics are built into the TorchScript system, such as tensor types.
+TorchScript defines the semantics of these builtin nominal types, and often support only a subset of the methods or
+attributes of its Python class definition.
+
+::
+
+    TSBuiltinClass := TSTensor | "torch.device" | "torch.Stream" | "torch.dtype" |
+                    "torch.nn.ModuleList" | "torch.nn.ModuleDict" | ...
+    TSTensor := "torch.Tensor" | "common.SubTensor" | "common.SubWithTorchFunction" |
+                "torch.nn.parameter.Parameter" | and subclasses of torch.Tensor
+
+
+Special note on torch.nn.ModuleList and torch.nn.ModuleDict
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+Although torch.nn.ModuleList and torch.nn.ModuleDict are defined as a list and dictionary in Python,
+in TorchScript, ``torch.nn.ModuleList`` and ``torch.nn.ModuleDict`` behave more like a tuple.
+
+    * Within the execution of TorchScript, instances of torch.nn.ModuleList  or torch.nn.ModuleDict is immutable.
+    * Iterating over torch.nn.ModuleList or torch.nn.ModuleDict is completely unrolled so that elements of torch.nn.ModuleList or keys of torch.nn.ModuleDict can be of different subclasses of torch.nn.Module.
+
+Example:
+
+.. testcode::
+
+    import torch
+
+    @torch.jit.script
+    class A():
+        def __init__(self):
+            self.x = torch.rand(3)
+
+        def f(self, y: torch.device):
+            return self.x.to(device=y)
+
+    def g():
+        a = A()
+        return a.f(torch.device("cpu"))
+
+    script_g = torch.jit.script(g)
+    print(script_g.graph)
+
+Custom Class
+^^^^^^^^^^^^
+
+Unlike built-in classes, semantics of custom classes are user-defined and the entire class definition must be compilable into TorchScript IR and subject to TorchScript type-checking rules.
+
+::
+
+    TSClassDef := [ "@torch.jit.script" ]
+                "class" ClassName [ "(object)" ]  ":"
+                        MethodDefinition
+
+where
+
+    * Classes must be new-style classes (note that Python 3 supports only new-style classes, for Python 2.x new-style class is specified by subclassing from object)
```

Comment:
PyTorch no longer supports Python 2, so not sure if this statement is still relevant.

Reply:
I agree.